### ISSUE TITLE
storage: leaseholder preferences are not enforced when the preferred replica is behind the quorum

### ISSUE THREAD

**Issue by awoods187** [cockroachdb/cockroach#38065]
A customer recently an into a problem with Follow the Workload. 

**Background:**
Cluster topology is 3 nodes in 3 separate regions
Most connections are going to Asia (using the right connection string) 
Determine this is due to the setup of the cluster and their use case
Admin UI is slow because customer is connecting to it through the global LB

Follow-the-workload didn’t seem to work:
- 80-90% of connections were issued in Asia but the leaseholders did not move there 
- The leases did not move as evidenced by the SQL CLI `SHOW RANGE`
- Localities were enabled.
- N1 (asia-southeast1) received nearly all connections & queries
- System.users and wallets table leases were on n2 (us-central1)
- Alter zone configs with leaseholder preferences return successfully but do not actually alter the zone configuration, and leases do not transfer.

**Comment by ajwerner** [comment]
I am not sure this is a follow the workload problem. There will always be a tension between optimizing for latency and balancing load. In this specific case there was an egregiously bad zone configuration. All of the indices for the table in question (as well as indices for some other tables) were set to have leaseholder preference in Asia however the table itself (and its primary index) did not have that lease holder preference set. The primary index for the table was placed in Europe. Traffic to that primary index represented roughly 100% of the traffic to the node in Europe. If follow-the-workload were to move the lease to Asia, it would lead the node in Europe with zero traffic while adding to the next-most-loaded node. I think I'd chalk this specific case up as user error.

**Comment by ajwerner** [comment]
> Alter zone configs with leaseholder preferences return successfully but do not actually alter the zone configuration, and leases do not transfer.

This I want to dig in to. What do you mean?

**Comment by awoods187** [comment]
@tim-o for the response on the alter zone

**Comment by tim-o** [comment]
```alter table <TableName> CONFIGURE ZONE USING lease_preferences = '[[+region=gcp-asia-southeast1]]';
CONFIGURE ZONE 1

Time: 1.011314805s```

Running `SHOW ZONE CONFIGURATIONS;` afterwards returns no result for <TableName>, even hours after the fact.

**Comment by ajwerner** [comment]
Okay that's the real bug here. Should we re-label the issue or create a new one?

**Comment by awoods187** [comment]
Let's relabel as you see fit @ajwerner

**Comment by vilterp** [comment]
I would say not being able to set a zone config on a table should be a new bug.

There still may be something about follow the workload not working; let's leave this bug to discuss that.

**Comment by awoods187** [comment]
Alright I opened https://github.com/cockroachdb/cockroach/issues/38074 for the zone config issue. @vilterp what makes you think this isn't user error?

**Comment by ajwerner** [comment]
I think we probably need a low severity issue to discuss the nuance between balance load and following the workload, we can use this for that. There should be a high severity issue for the `ALTER TABLE ZONE ... CONFIGURE`. @tim-o you're the one who experienced it, any chance you could write up your experience? It seems like there's a related bug where we cannot load the Data Distribute & Zone Configuration page.

**Comment by ajwerner** [comment]
We now understand why the lease isn't being transferred better and it's pretty interesting.

The cluster here has 3 nodes, one each in Asia (n1), Europe (n2), and the US (n3). See the latency chart:

![Screenshot_2019-06-06 Network Diagnostics Debug Cockroach Console](https://user-images.githubusercontent.com/1839234/59068657-b5478500-8882-11e9-82af-fd51288bfd5b.png)

Most of the traffic is from Asia and furthermore we've set leaseholder preferences to be in Asia, yet we see that the leaseholder for the primary index is stuck in Europe. Why? Shouldn't the system transfer it? Well, yes, we do want to transfer the lease but unfortunately the we're foiled by this [strict logic to prune behind replicas](https://github.com/cockroachdb/cockroach/blob/e860447930698a46264d42ec943d2c92a1ae22fd/pkg/storage/allocator.go#L1194) and the speed of light. In this topology, under this constant workload, it is roughly never the case that the faraway node won't be behind. This logic is very strict, much stricter than the logic in the quotaPool which allows replicas which are making progress but a bit behind due to replication latency to not prevent writes (see [here](https://github.com/cockroachdb/cockroach/blob/e860447930698a46264d42ec943d2c92a1ae22fd/pkg/storage/replica_proposal_quota.go#L158)).

**Comment by darinpp** [comment]
This is fixed in master and the fix is backported to 19.2/19.1
